test/e2e/framework/cluster.go

```diff
@@ -0,0 +1,330 @@
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ServiceAccountTokenSecretName,
+			Annotations: map[string]string{
+				corev1.ServiceAccountNameKey: userSA.Name,
+			},
+		},
+		Type: corev1.SecretTypeServiceAccountToken,
+	}, metav1.CreateOptions{})
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	By("Waiting for service account token Secret %q in namespace %q.", userSATokenSecret.Name, userSATokenSecret.Namespace)
+	ctxUserSATokenSecret, ctxUserSATokenSecretCancel := context.WithTimeout(ctx, serviceAccountTokenSecretWaitTimeout)
+	defer ctxUserSATokenSecretCancel()
+	userSATokenSecret, err = WaitForServiceAccountTokenSecret(ctxUserSATokenSecret, c.KubeAdminClient().CoreV1(), userSATokenSecret.Namespace, userSATokenSecret.Name)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	o.Expect(userSATokenSecret.Data).To(o.HaveKey(corev1.ServiceAccountTokenKey))
+
+	token := userSATokenSecret.Data[corev1.ServiceAccountTokenKey]
+	o.Expect(token).NotTo(o.BeEmpty())
+
+	// Create a restricted client using the user SA.
+	c.clientConfig = restclient.AnonymousClientConfig(c.AdminClientConfig())
+	c.clientConfig.BearerToken = string(token)
+
+	// Wait for default ServiceAccount.
+	By("Waiting for default ServiceAccount in namespace %q.", ns.Name)
+	ctxSa, ctxSaCancel := context.WithTimeout(ctx, serviceAccountWaitTimeout)
+	defer ctxSaCancel()
+	_, err = WaitForServiceAccount(ctxSa, c.KubeAdminClient().CoreV1(), ns.Name, "default")
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	// Waits for the configmap kube-root-ca.crt containing CA trust bundle so that pods do not have to retry mounting
+	// the config map because it creates noise and slows the startup.
+	By("Waiting for kube-root-ca.crt in namespace %q.", ns.Name)
+	_, err = controllerhelpers.WaitForConfigMapState(
+		ctx,
+		c.KubeClient().CoreV1().ConfigMaps(ns.Name),
+		"kube-root-ca.crt",
+		controllerhelpers.WaitForStateOptions{},
+		func(configMap *corev1.ConfigMap) (bool, error) {
+			return true, nil
+		},
+	)
+	o.Expect(err).NotTo(o.HaveOccurred())
+}
+
+func (c *Cluster) deleteNamespace(ctx context.Context, ns *corev1.Namespace) {
+	By("Destroying namespace %q.", ns.Name)
+	var gracePeriod int64 = 0
+	var propagation = metav1.DeletePropagationForeground
+	err := c.KubeAdminClient().CoreV1().Namespaces().Delete(
+		ctx,
+		ns.Name,
+		metav1.DeleteOptions{
+			GracePeriodSeconds: &gracePeriod,
+			PropagationPolicy:  &propagation,
+			Preconditions: &metav1.Preconditions{
+				UID: &ns.UID,
+			},
+		},
+	)
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	// We have deleted only the namespace object but it is still there with deletionTimestamp set.
+
+	By("Waiting for namespace %q to be removed.", ns.Name)
+	err = WaitForObjectDeletion(ctx, c.DynamicAdminClient(), corev1.SchemeGroupVersion.WithResource("namespaces"), "", ns.Name, &ns.UID)
+	o.Expect(err).NotTo(o.HaveOccurred())
+	klog.InfoS("Namespace removed.", "Namespace", ns.Name)
+}
+
+func (c *Cluster) beforeEach(ctx context.Context) {
```

Comment:
I'd expect these on the framework level - make a namespace per `It` on the cluster 0, the rest is on demand

Reply:
that's not going to sit well with the cluster interface - there's a `func Namespace() string`, which returns the auto namespace and adding an error to it would require changing a lot of existing code. I don't really see this as part of framework struct. `setupNamespace` also configures the clientConfig, so `KubeClient` and similar functions would have to become optional as well? We could wrap them for the framework, but then having the framework implement the cluster interface doesn't really work without changes to the existing test. Why not have it set up for every cluster anyway?

Reply:
I am fine with having the `Namespace()` there for manual call, just not the beforEach in all of them. that seems easily doable by having the before each set only on the cluster and calling it on the cluster index 0.

Reply:
Having the namespace there is not the issue, it's just that it can't return an error without changing the code of all our tests. I added a safeguard though - best I can do here. 
I changed the framework to only set up and destroy namespace for the default cluster.